builder/vsphere/driver/vm.go

```diff
@@ -339,6 +347,115 @@ func (vm *VirtualMachine) Clone(ctx context.Context, config *CloneConfig) (*Virt
 	return created, nil
 }
 
+func (vm *VirtualMachine) updateVAppConfig(newProps map[string]string) (*types.VmConfigSpec, error) {
+	if len(newProps) == 0 {
+		return nil, nil
+	}
+
+	vProps, _ := vm.Properties()
+	if vProps.Config.VAppConfig == nil {
+		return nil, fmt.Errorf("this VM lacks a vApp configuration and cannot have vApp properties set on it")
+	}
+
+	allProperties := vProps.Config.VAppConfig.GetVmConfigInfo().Property
+
+	var props []types.VAppPropertySpec
+	for _, p := range allProperties {
+		if *p.UserConfigurable == true {
+			defaultValue := " "
+			if p.DefaultValue != "" {
+				defaultValue = p.DefaultValue
+			}
+			prop := types.VAppPropertySpec{
+				ArrayUpdateSpec: types.ArrayUpdateSpec{
+					Operation: types.ArrayUpdateOperationEdit,
+				},
+				Info: &types.VAppPropertyInfo{
+					Key:              p.Key,
+					Id:               p.Id,
+					Value:            defaultValue,
+					UserConfigurable: p.UserConfigurable,
+				},
+			}
+
+			newValue, ok := newProps[p.Id]
+			if ok {
+				prop.Info.Value = newValue
+				delete(newProps, p.Id)
+			}
+			props = append(props, prop)
+		} else {
+			_, ok := newProps[p.Id]
+			if ok {
+				return nil, fmt.Errorf("vApp property with userConfigurable=false specified in vapp.properties: %+v", reflect.ValueOf(newProps).MapKeys())
+			}
+		}
+	}
+
+	if len(newProps) > 0 {
+		return nil, fmt.Errorf("unsupported vApp properties in vapp.properties: %+v", reflect.ValueOf(newProps).MapKeys())
+	}
+
+	return &types.VmConfigSpec{
+		Property: props,
+	}, nil
+}
+
+func (vm *VirtualMachine) AddPublicKeys(publicKeys string) error {
```

Comment:
Is this called ? I can't see eye 👀 . Also may be this could simply be an addition to the `updateVAppConfig` call to avoid repetition ?

Reply:
This is used by the `StepSshKeyPair` line 85. 
I can use the updateVAppConfig now that I've changed with your suggestion above. Before that, the `updateVAppConfig` was updating all of the properties and that wasn't necessary for this case, but now I've changed it to update only with what's in the newProps map. So now I can just pass the the "public-keys" in the map and it will work the same way 👍